File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ChatFromViewer.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * ChatFromViewer
 * Specifies the text to be said and the "type",
 * normal speech, shout, whisper.
 * with the specified radius
 *
 * <p>Template: {@code ChatFromViewer Low 80 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class ChatFromViewer extends SLMessage {
    public AgentData AgentData_Field;
    public ChatData ChatData_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block ChatData, Single. */
    public static class ChatData {
        public int Channel; // S32
        public byte[] Message; // Variable 2
        public int Type; // U8
    }

    public ChatFromViewer() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.ChatData_Field = new ChatData();
    }

    @Override
    public int CalcPayloadSize() {
        return this.ChatData_Field.Message.length + 2 + 1 + 4 + 36;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleChatFromViewer(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 80 (ChatFromViewer).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x50);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packVariable(byteBuffer, this.ChatData_Field.Message, 2);
        packByte(byteBuffer, (byte) this.ChatData_Field.Type);
        packInt(byteBuffer, this.ChatData_Field.Channel);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.ChatData_Field.Message = unpackVariable(byteBuffer, 2);
        this.ChatData_Field.Type = unpackByte(byteBuffer) & 0xFF;
        this.ChatData_Field.Channel = unpackInt(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ChatFromViewer.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * ChatFromViewer
 * Specifies the text to be said and the "type",
 * normal speech, shout, whisper.
 * with the specified radius
 *
 * <p>Template: {@code ChatFromViewer Low 80 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class ChatFromViewer : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var ChatData_Field: ChatData = ChatData()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block ChatData, Single. */
    open class ChatData {
        @JvmField var Channel: Int = 0
        @JvmField var Message: ByteArray? = null
        @JvmField var Type: Int = 0
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return ChatData_Field.Message!!.size + 2 + 1 + 4 + 36
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleChatFromViewer(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 80 (ChatFromViewer).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x50).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packVariable(byteBuffer, ChatData_Field.Message, 2)
        packByte(byteBuffer, (ChatData_Field.Type).toByte())
        packInt(byteBuffer, ChatData_Field.Channel)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        ChatData_Field.Message = unpackVariable(byteBuffer, 2)
        ChatData_Field.Type = unpackByte(byteBuffer).toInt() and 0xFF
        ChatData_Field.Channel = unpackInt(byteBuffer)
    }
}
